src/prelude/internals.mo

```diff
@@ -589,30 +589,33 @@ func @timer_helper() : async () {
     case null ();
     case (?n) {
       gatherExpired(n.pre);
-      if (n.expire[0] > 0 and n.expire[0] <= now and gathered < thunks.size()) {
-        thunks[gathered] := ?(n.job);
-        switch (n.delay) {
-          case (null or ?0) ();
-          case (?delay) {
-            // re-add the node, skipping past expirations
-            let expire = n.expire[0] + delay * (1 + (now - n.expire[0]) / delay);
-            n.expire[0] := 0;
-            // N.B. reinsert only works on pruned nodes
-            func reinsert(m : ?@Node) : @Node = switch m {
-              case null ({ n with expire = [var expire]; pre = null; post = null });
-              case (?m) {
-                assert m.expire[0] != 0;
-                if (expire < m.expire[0]) ({ m with pre = ?reinsert(m.pre) })
-                else ({ m with post = ?reinsert(m.post) })
-              }
-            };
-            @timers := ?reinsert(@prune(@timers));
+      if (n.expire[0] <= now and gathered < thunks.size()) {
+        if (n.expire[0] > 0) {
+          // not expunged yet
+          thunks[gathered] := ?(n.job);
+          switch (n.delay) {
+            case (null or ?0) ();
+            case (?delay) {
+              // re-add the node, skipping past expirations
+              let expire = n.expire[0] + delay * (1 + (now - n.expire[0]) / delay);
+              n.expire[0] := 0;
+              // N.B. reinsert only works on pruned nodes
+              func reinsert(m : ?@Node) : @Node = switch m {
+                case null ({ n with expire = [var expire]; pre = null; post = null });
+                case (?m) {
+                  assert m.expire[0] != 0;
+                  if (expire < m.expire[0]) ({ m with pre = ?reinsert(m.pre) })
+                  else ({ m with post = ?reinsert(m.post) })
+                }
+              };
+              @timers := ?reinsert(@prune(@timers));
+            }
           };
+          n.expire[0] := 0;
+          gathered += 1;
         };
-        n.expire[0] := 0;
-        gathered += 1;
-      };
-      gatherExpired(n.post);
+        gatherExpired(n.post)
```

Comment:
In contrast to prior practice (where we traversed the `post` part unconditionally), we now only attempt to gather from the`post` nodes when
- there are still slots to fill, and
- expiration of the currently visited node is _not_ in the future (i.e. either past/present or expunged).